File: recamera-intellisense-mcp/src/capture.rs
use anyhow::{bail, Result};
use serde_json::{json, Value};

use crate::api_client::ApiClient;
use crate::types::*;

const CAPTURE_OUTPUT_DEFAULT: &str = "/mnt/rc_mmcblk0p8/reCamera";
const CAPTURE_FORMAT_IMAGE: &str = "JPG";
const CAPTURE_POLL_INTERVAL_MS: u64 = 500;
const CAPTURE_TIMEOUT_SECS: u64 = 5;

fn parse_capture_event(data: &Value) -> CaptureEvent {
    CaptureEvent {
        id: data
            .get("sID")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        output_directory: data
            .get("sOutputDirectory")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        format: data
            .get("sFormat")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        video_length_seconds: data
            .get("iVideoLengthSeconds")
            .and_then(|v| v.as_i64())
            .map(|v| v as i32),
        status: data
            .get("sStatus")
            .and_then(|v| v.as_str())
            .unwrap_or("UNKNOWN")
            .to_string(),
        timestamp_unix_ms: data.get("iTimestamp").and_then(|v| v.as_u64()).unwrap_or(0),
        file_name: data
            .get("sFileName")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
    }
}

pub async fn get_capture_status(
    client: &ApiClient,
    device: &DeviceRecord,
) -> Result<CaptureStatus> {
    let data = client
        .get_json(device, "/cgi-bin/entry.cgi/record/capture/status", None)
        .await?;
    let last_capture = data
        .get("dLastCapture")
        .filter(|v| v.is_object())
        .map(parse_capture_event);
    Ok(CaptureStatus {
        last_capture,
        ready_to_start_new: data
            .get("bReadyToStartNew")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
        stop_requested: data
            .get("bStopRequested")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
    })
}

pub async fn start_capture(
    client: &ApiClient,
    device: &DeviceRecord,
    output: Option<&str>,
    format: Option<&str>,
    video_length_seconds: Option<i32>,
) -> Result<CaptureEvent> {
    let mut payload = json!({
        "sOutput": output.unwrap_or(CAPTURE_OUTPUT_DEFAULT),
        "sFormat": format.unwrap_or(CAPTURE_FORMAT_IMAGE).to_uppercase(),
    });
    if let Some(vl) = video_length_seconds {
        payload["iVideoLengthSeconds"] = json!(vl);
    }
    let result = client
        .post_json(
            device,
            "/cgi-bin/entry.cgi/record/capture/start",
            None,
            Some(&payload),
        )
        .await?;
    if result.get("code").and_then(|v| v.as_i64()).unwrap_or(-1) != 0 {
        let msg = result
            .get("message")
            .and_then(|v| v.as_str())
            .unwrap_or("Unknown error");
        bail!("Failed to start capture: {msg}");
    }
    let capture_data = result
        .get("dCapture")
        .ok_or_else(|| anyhow::anyhow!("Missing capture event in response"))?;
    Ok(parse_capture_event(capture_data))
}

pub async fn stop_capture(client: &ApiClient, device: &DeviceRecord) -> Result<()> {
    let result = client
        .post_json(device, "/cgi-bin/entry.cgi/record/capture/stop", None, None)
        .await?;
    if result.get("code").and_then(|v| v.as_i64()).unwrap_or(-1) != 0 {
        let msg = result
            .get("message")
            .and_then(|v| v.as_str())
            .unwrap_or("Unknown error");
        bail!("Failed to stop capture: {msg}");
    }
    Ok(())
}

const TERMINAL_STATUSES: &[&str] = &["COMPLETED", "FAILED", "INTERRUPTED", "CANCELED"];

/// Capture an image, poll for completion, and download it.
/// Returns (CaptureEvent, image_bytes).
pub async fn capture_image(
    client: &ApiClient,
    device: &DeviceRecord,
    output: Option<&str>,
    timeout_seconds: Option<u64>,
) -> Result<(CaptureEvent, Vec<u8>)> {
    // If no output is specified, get the default enabled storage and use its path as output
    let output = if let Some(output) = output {
        output.to_string()
    } else {
        let storages = crate::storage::get_storage_list(client, device).await?;
        storages
            .iter()
            .find(|s| s.is_enabled)
            .filter(|s| !s.mount_path.is_empty())
            .and_then(|s| {
                let full_path = std::path::Path::new(&s.mount_path).join(&s.data_dir);
                full_path.to_str().map(|p| p.to_string())
            })
            .unwrap_or_else(|| CAPTURE_OUTPUT_DEFAULT.to_string())
    };
    let capture = start_capture(
        client,
        device,
        Some(output.as_str()),
        Some(CAPTURE_FORMAT_IMAGE),
        None,
    )
    .await?;
    let timeout = timeout_seconds.unwrap_or(CAPTURE_TIMEOUT_SECS);
    let interval = std::time::Duration::from_millis(CAPTURE_POLL_INTERVAL_MS);
    let deadline = tokio::time::Instant::now() + tokio::time::Duration::from_secs(timeout);
    let mut final_capture = capture.clone();
    while tokio::time::Instant::now() < deadline {
        let status = get_capture_status(client, device).await?;
        if let Some(last) = &status.last_capture {
            tokio::time::sleep(interval).await;
            if last.id == capture.id && TERMINAL_STATUSES.contains(&last.status.as_str()) {
                final_capture = last.clone();
                break;
            }
        }
    }
    if final_capture.status != "COMPLETED" {
        bail!(
            "Capture did not complete successfully (status: '{}')",
            final_capture.status
        );
    }
    let remote_path = format!(
        "{}/{}",
        final_capture.output_directory, final_capture.file_name
    );
    let image_data = crate::storage::fetch_file(client, device, &remote_path).await?;
    Ok((final_capture, image_data))
}

File: skills/recamera-local/scripts/recamera_local/capture.py
"""Capture status, start/stop, and the `capture_image` helper (local edition).

Captures land on the local filesystem, so `capture_image` returns the
**path** by default — the agent opens it with its own file/vision tooling.
Base64 is opt-in (`inline=true`) for agents without file tools.
"""

from __future__ import annotations

import base64
import os
import time
from typing import Any, Dict, Optional

from . import _local as http
from ._coerce import to_bool
from ._const import MAX_INLINE_BYTES
from ._errors import RecameraError

__all__ = ["get_capture_status", "start_capture", "stop_capture", "capture_image"]

PATH_STATUS = "/cgi-bin/entry.cgi/record/capture/status"
PATH_START = "/cgi-bin/entry.cgi/record/capture/start"
PATH_STOP = "/cgi-bin/entry.cgi/record/capture/stop"

FORMAT_IMAGE = "JPG"
OUTPUT_FALLBACK = "/mnt/rc_mmcblk0p8/DCIM/100RECAM"
_POLL_INTERVAL_S = 0.25
_DEFAULT_TIMEOUT_S = 5.0
_TERMINAL = {"COMPLETED", "FAILED", "INTERRUPTED", "CANCELED"}


def _parse_event(d: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "id": d.get("sID", ""),
        "output_directory": d.get("sOutputDirectory", ""),
        "format": d.get("sFormat", ""),
        "video_length_seconds": d.get("iVideoLengthSeconds"),
        "status": d.get("sStatus", "UNKNOWN"),
        "timestamp_unix_ms": int(d.get("iTimestamp", 0) or 0),
        "file_name": d.get("sFileName", ""),
    }


def get_capture_status() -> Dict[str, Any]:
    """Current capture state (includes the last event, if any)."""
    d = http.get_json(PATH_STATUS) or {}
    last = d.get("dLastCapture")
    return {
        "last_capture": _parse_event(last) if isinstance(last, dict) else None,
        "ready_to_start_new": bool(d.get("bReadyToStartNew", False)),
        "stop_requested": bool(d.get("bStopRequested", False)),
    }


def _validate_output_dir(output: Optional[str]) -> Optional[str]:
    """`output` is an absolute directory on the *camera's* filesystem."""
    if output is None:
        return None
    if not isinstance(output, str) or not output.strip():
        raise ValueError("'output' must be a non-empty string.")
    if "\x00" in output:
        raise ValueError("'output' must not contain NUL bytes.")
    if not output.startswith("/"):
        raise ValueError(
            f"'output' must be an absolute on-device directory path; got {output!r}. "
            "Use get_storage_status to find a mount path (e.g. "
            "'/mnt/rc_mmcblk0p8/DCIM/100RECAM'), or omit 'output' to use the "
            "selected storage slot."
        )
    return output


def start_capture(
    *,
    output: Optional[str] = None,
    format: str = FORMAT_IMAGE,
    video_length_seconds: Optional[int] = None,
) -> Dict[str, Any]:
    """Start a capture; returns the initial capture event.

    `output` is an on-device directory. Omit it to auto-resolve from the
    currently selected storage slot. The device picks the file name; the
    resulting file is `event.output_directory + '/' + event.file_name` —
    readable directly once the event reaches a terminal status.
    """
    output = _validate_output_dir(output)
    payload: Dict[str, Any] = {
        "sOutput": output or OUTPUT_FALLBACK,
        "sFormat": str(format).upper(),
    }
    if video_length_seconds is not None:
        payload["iVideoLengthSeconds"] = int(video_length_seconds)
    try:
        resp = http.post_json(PATH_START, payload=payload)
        http.expect_ok(resp, "start capture")
    except RecameraError as exc:
        # Device rejects paths outside a storage mount (error code 30022).
        if exc.code == 30022 or "30022" in str(exc):
            raise RecameraError(
                f"{exc} Hint: 'output' must be an on-device directory under a "
                "mounted storage slot (see get_storage_status for mount_path), "
                "or omit 'output' to use the default.",
                status=exc.status,
                code=exc.code,
                body=exc.body,
            ) from exc
        raise
    capture = resp.get("dCapture")
    if not isinstance(capture, dict):
        raise RecameraError("start_capture response missing dCapture field.")
    return _parse_event(capture)


def stop_capture() -> None:
    """Stop the running capture (no-op for JPG)."""
    resp = http.post_json(PATH_STOP)
    http.expect_ok(resp, "stop capture")


def capture_image(
    *,
    output: Optional[str] = None,
    timeout: float = _DEFAULT_TIMEOUT_S,
    inline: bool = False,
    max_inline_bytes: int = MAX_INLINE_BYTES,
) -> Dict[str, Any]:
    """One-shot JPG: start, wait for completion, return `{event, path, size}`.

    With `inline=true`, also embeds `content_base64` (skipped with a `note`
    when the file exceeds `max_inline_bytes`).
    """
    inline = to_bool(inline, "inline")
    if output is None:
        try:
            from .storage import get_storage_status

            slots = get_storage_status()
            slot = next((s for s in slots if s["enabled"] and s["mount_path"]), None)
            if slot:
                base = slot["mount_path"].rstrip("/")
                data_dir = slot.get("data_dir", "").strip("/")
                output = f"{base}/{data_dir}" if data_dir else base
        except RecameraError:
            output = None
    output = output or OUTPUT_FALLBACK

    capture = start_capture(output=output, format=FORMAT_IMAGE)
    deadline = time.time() + float(timeout)
    final = dict(capture)
    while time.time() < deadline:
        time.sleep(_POLL_INTERVAL_S)
        status = get_capture_status()
        last = status["last_capture"]
        if last and last["id"] == capture["id"] and last["status"] in _TERMINAL:
            final = last
            break
    if final["status"] != "COMPLETED":
        raise RecameraError(f"Capture did not complete (status: {final['status']!r})")

    remote = f"{final['output_directory'].rstrip('/')}/{final['file_name']}"
    try:
        size = os.path.getsize(remote)
    except OSError as exc:
        raise RecameraError(
            f"Capture completed but {remote!r} is not readable locally: {exc}"
        ) from exc

    result: Dict[str, Any] = {"event": final, "path": remote, "size": size}
    if inline:
        if size <= int(max_inline_bytes):
            with open(remote, "rb") as fh:
                result["content_base64"] = base64.b64encode(fh.read()).decode("ascii")
        else:
            result["note"] = (
                f"file is {size} bytes, over the inline budget "
                f"({int(max_inline_bytes)}); open {remote!r} directly instead."
            )
    return result


COMMANDS = {
    "get_capture_status": get_capture_status,
    "start_capture": start_capture,
    "stop_capture": stop_capture,
    "capture_image": capture_image,
}
